### PR TITLE
Fix logical error "Expected ReadBufferFromFile, but got DB::EmptyReadBuffer"

### LINKED ISSUE #65877: Expected ReadBufferFromFile, but got DB::EmptyReadBuffer
```
2024.06.30 20:15:42.857905 [ 2417 ] {9c733732-76a1-4fc5-81e2-643cd76721f9::2_2_2_1_4} <Fatal> : Logical error: 'Expected ReadBufferFromFile, but got DB::EmptyReadBuffer'.
2024.06.30 20:50:10.124384 [ 50498 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.06.30 20:50:10.124668 [ 50498 ] {} <Fatal> BaseDaemon: (version 24.7.1.656, build id: 75E1A5225127BB742190A3760D8AB4BBFC75B1C2, git hash: 403e062d4ff62c3b155528ba35bff5963c31aee4) (from thread 2417) Received signal 6
2024.06.30 20:50:10.124905 [ 50498 ] {} <Fatal> BaseDaemon: Signal description: Aborted
2024.06.30 20:50:10.125037 [ 50498 ] {} <Fatal> BaseDaemon: 
2024.06.30 20:50:10.125227 [ 50498 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000001388a6aa 0x000000001388a5a0 0x0000000013cbb151 0x00007f7002d04520 0x00007f7002d589fd 0x00007f7002d04476 0x00007f7002cea7f3 0x000000001382ae82 0x000000001382aff5 0x000000001382b5a0 0x000000000ab6220a 0x000000000ab5f9e9 0x000000000ab5f7e5 0x000000001cff18fd 0x000000001cff6398 0x000000001cff5e92 0x000000001d5ce900 0x000000001d589824 0x000000001d5896ef 0x000000001d58f8ef 0x000000001c059385 0x000000001bfe8dd7 0x000000001bfe32ca 0x000000001d9366c7 0x000000001d94ba3f 0x0000000022cbe299 0x0000000022cbeae8 0x0000000022c49c8e 0x0000000022c4641a 0x0000000022c4515e 0x00007f7002d56ac3 0x00007f7002de8850
2024.06.30 20:50:10.125434 [ 50498 ] {} <Fatal> BaseDaemon: ########################################
2024.06.30 20:50:10.125774 [ 50498 ] {} <Fatal> BaseDaemon: (version 24.7.1.656, build id: 75E1A5225127BB742190A3760D8AB4BBFC75B1C2, git hash: 403e062d4ff62c3b155528ba35bff5963c31aee4) (from thread 2417) (query_id: 9c733732-76a1-4fc5-81e2-643cd76721f9::2_2_2_1_4) (query: ) Received signal Aborted (6)
2024.06.30 20:50:10.126034 [ 50498 ] {} <Fatal> BaseDaemon: 
2024.06.30 20:50:10.126135 [ 50498 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000001388a6aa 0x000000001388a5a0 0x0000000013cbb151 0x00007f7002d04520 0x00007f7002d589fd 0x00007f7002d04476 0x00007f7002cea7f3 0x000000001382ae82 0x000000001382aff5 0x000000001382b5a0 0x000000000ab6220a 0x000000000ab5f9e9 0x000000000ab5f7e5 0x000000001cff18fd 0x000000001cff6398 0x000000001cff5e92 0x000000001d5ce900 0x000000001d589824 0x000000001d5896ef 0x000000001d58f8ef 0x000000001c059385 0x000000001bfe8dd7 0x000000001bfe32ca 0x000000001d9366c7 0x000000001d94ba3f 0x0000000022cbe299 0x0000000022cbeae8 0x0000000022c49c8e 0x0000000022c4641a 0x0000000022c4515e 0x00007f7002d56ac3 0x00007f7002de8850
2024.06.30 20:50:10.483055 [ 50498 ] {} <Fatal> BaseDaemon: 0. /build/src/Common/StackTrace.cpp:349: StackTrace::tryCapture() @ 0x000000001388a6aa
2024.06.30 20:50:10.676125 [ 50498 ] {} <Fatal> BaseDaemon: 1. /build/src/Common/StackTrace.cpp:323: StackTrace::StackTrace(ucontext_t const&) @ 0x000000001388a5a0
2024.06.30 20:50:11.022430 [ 50498 ] {} <Fatal> BaseDaemon: 2. /build/src/Daemon/BaseDaemon.cpp:158: signalHandler(int, siginfo_t*, void*) @ 0x0000000013cbb151
2024.06.30 20:50:11.022684 [ 50498 ] {} <Fatal> BaseDaemon: 3. ? @ 0x00007f7002d04520
2024.06.30 20:50:11.022830 [ 50498 ] {} <Fatal> BaseDaemon: 4. ? @ 0x00007f7002d589fd
2024.06.30 20:50:11.022979 [ 50498 ] {} <Fatal> BaseDaemon: 5. ? @ 0x00007f7002d04476
2024.06.30 20:50:11.023118 [ 50498 ] {} <Fatal> BaseDaemon: 6. ? @ 0x00007f7002cea7f3
2024.06.30 20:50:11.257525 [ 50498 ] {} <Fatal> BaseDaemon: 7. /build/src/Common/Exception.cpp:0: DB::abortOnFailedAssertion(String const&) @ 0x000000001382ae82
2024.06.30 20:50:11.506572 [ 50498 ] {} <Fatal> BaseDaemon: 8. /build/src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&) @ 0x000000001382aff5
2024.06.30 20:50:11.751451 [ 50498 ] {} <Fatal> BaseDaemon: 9. /build/src/Common/Exception.cpp:105: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000001382b5a0
2024.06.30 20:50:11.899275 [ 50498 ] {} <Fatal> BaseDaemon: 10. /build/src/Common/Exception.h:95: DB::Exception::Exception(String&&, int, bool) @ 0x000000000ab6220a
2024.06.30 20:50:12.097911 [ 50498 ] {} <Fatal> BaseDaemon: 11. /build/src/Common/Exception.h:68: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000ab5f9e9
2024.06.30 20:50:12.277556 [ 50498 ] {} <Fatal> BaseDaemon: 12. /build/src/Common/Exception.h:113: DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x000000000ab5f7e5
2024.06.30 20:50:13.185144 [ 50498 ] {} <Fatal> BaseDaemon: 13. /build/src/Storages/MergeTree/MergeTask.cpp:562: DB::MergeTask::VerticalMergeStage::prepareVerticalMergeForAllColumns() const @ 0x000000001cff18fd
2024.06.30 20:50:13.950082 [ 50498 ] {} <Fatal> BaseDaemon: 14. /build/src/Storages/MergeTree/MergeTask.cpp:891: DB::MergeTask::VerticalMergeStage::execute() @ 0x000000001cff6398
2024.06.30 20:50:14.750855 [ 50498 ] {} <Fatal> BaseDaemon: 15. /build/src/Storages/MergeTree/MergeTask.cpp:951: DB::MergeTask::execute() @ 0x000000001cff5e92
2024.06.30 20:50:14.892867 [ 50498 ] {} <Fatal> BaseDaemon: 16. /build/src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:57: DB::MergePlainMergeTreeTask::executeStep() @ 0x000000001d5ce900
2024.06.30 20:50:15.787768 [ 50498 ] {} <Fatal> BaseDaemon: 17. /build/src/Storages/MergeTree/MergePlainMergeTreeTask.h:104: DB::executeHere(std::shared_ptr<DB::MergePlainMergeTreeTask>) @ 0x000000001d589824
2024.06.30 20:50:16.500116 [ 50498 ] {} <Fatal> BaseDaemon: 18. /build/src/Storages/StorageMergeTree.cpp:1170: DB::StorageMergeTree::merge(bool, String const&, bool, bool, std::vector<String, std::allocator<String>> const&, bool, std::shared_ptr<DB::MergeTreeTransaction> const&, PreformattedMessage&, bool) @ 0x000000001d5896ef
2024.06.30 20:50:17.129328 [ 50498 ] {} <Fatal> BaseDaemon: 19. /build/src/Storages/StorageMergeTree.cpp:1582: DB::StorageMergeTree::optimize(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::IAST> const&, bool, bool, std::vector<String, std::allocator<String>> const&, bool, std::shared_ptr<DB::Context const>) @ 0x000000001d58f8ef
2024.06.30 20:50:17.232362 [ 50498 ] {} <Fatal> BaseDaemon: 20. /build/src/Interpreters/InterpreterOptimizeQuery.cpp:84: DB::InterpreterOptimizeQuery::execute() @ 0x000000001c059385
2024.06.30 20:50:17.655183 [ 50498 ] {} <Fatal> BaseDaemon: 21. /build/src/Interpreters/executeQuery.cpp:1225: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001bfe8dd7
2024.06.30 20:50:18.105156 [ 50498 ] {} <Fatal> BaseDaemon: 22. /build/src/Interpreters/executeQuery.cpp:1391: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x000000001bfe32ca
2024.06.30 20:50:18.595800 [ 50498 ] {} <Fatal> BaseDaemon: 23. /build/src/Server/TCPHandler.cpp:521: DB::TCPHandler::runImpl() @ 0x000000001d9366c7
2024.06.30 20:50:19.135275 [ 50498 ] {} <Fatal> BaseDaemon: 24. /build/src/Server/TCPHandler.cpp:2353: DB::TCPHandler::run() @ 0x000000001d94ba3f
2024.06.30 20:50:19.169984 [ 50498 ] {} <Fatal> BaseDaemon: 25. /build/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x0000000022cbe299
2024.06.30 20:50:19.221942 [ 50498 ] {} <Fatal> BaseDaemon: 26. /build/base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x0000000022cbeae8
2024.06.30 20:50:19.431616 [ 50498 ] {} <Fatal> BaseDaemon: 27. /build/base/poco/Foundation/src/ThreadPool.cpp:205: Poco::PooledThread::run() @ 0x0000000022c49c8e
2024.06.30 20:50:19.484743 [ 50498 ] {} <Fatal> BaseDaemon: 28. /build/base/poco/Foundation/src/Thread.cpp:46: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x0000000022c4641a
2024.06.30 20:50:19.534619 [ 50498 ] {} <Fatal> BaseDaemon: 29. /build/base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000022c4515e
2024.06.30 20:50:19.535514 [ 50498 ] {} <Fatal> BaseDaemon: 30. ? @ 0x00007f7002d56ac3
2024.06.30 20:50:19.535632 [ 50498 ] {} <Fatal> BaseDaemon: 31. ? @ 0x00007f7002de8850
2024.06.30 20:50:19.535745 [ 50498 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.06.30 20:50:26.610739 [ 50498 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
2024.06.30 20:50:26.611218 [ 50498 ] {} <Fatal> BaseDaemon: Changed settings: connect_timeout_with_failover_ms = 2000, connect_timeout_with_failover_secure_ms = 3000, idle_connection_timeout = 36000, s3_check_objects_after_upload = true, stream_like_engine_allow_direct_select = true, replication_wait_for_inactive_replica_timeout = 30, allow_nonconst_timezone_arguments = true, log_queries = true, insert_quorum_timeout = 60000, fsync_metadata = false, http_send_timeout = 60., http_receive_timeout = 60., opentelemetry_start_trace_probability = 0.10000000149011612, max_untracked_memory = 1048576, memory_profiler_step = 1048576, allow_introspection_functions = true, database_atomic_wait_for_drop_and_detach_synchronously = true, distributed_ddl_entry_format_version = 6, async_insert_busy_timeout_max_ms = 5000, enable_filesystem_cache = true, enable_filesystem_cache_on_write_operations = true, filesystem_cache_segments_batch_size = 10, load_marks_asynchronously = true, allow_prefetched_read_pool_for_remote_filesystem = false, allow_prefetched_read_pool_for_local_filesystem = false, filesystem_prefetch_max_memory_usage = 1073741824, workload = 'default', insert_keeper_max_retries = 100, insert_keeper_retry_initial_backoff_ms = 1, insert_keeper_retry_max_backoff_ms = 1, insert_keeper_fault_injection_probability = 0.009999999776482582
2024.06.30 20:50:30.523429 [ 924 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

CI: https://s3.amazonaws.com/clickhouse-test-reports/64955/e61e13c3239ba17e8c9b0bd057c186bb9c080c59/stateless_tests__debug__s3_storage__[6_6].html
CI: https://s3.amazonaws.com/clickhouse-test-reports/64955/e61e13c3239ba17e8c9b0bd057c186bb9c080c59/stateless_tests__tsan__s3_storage__[1_5].html

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes https://github.com/ClickHouse/ClickHouse/issues/65877.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Review by antaljanosbenjamin — APPROVED** [review]

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit d4312fe54b508c2f69fff15d724a16881e235f93 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65949/d4312fe54b508c2f69fff15d724a16881e235f93/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/1e549751c703912a72eb7754c0a82fc1ddcdc19e/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/Cache/WriteBufferToFileSegment.cpp [hunk 1]
```diff
@@ -134,7 +134,7 @@ std::unique_ptr<ReadBuffer> WriteBufferToFileSegment::getReadBufferImpl()
     if (file_segment->getDownloadedSize() > 0)
         return std::make_unique<ReadBufferFromFile>(file_segment->getPath());
     else
-        return std::make_unique<EmptyReadBuffer>();
+        return std::make_unique<ReadBufferFromEmptyFile>();
```
Comment (Algunenano) [inline src/Interpreters/Cache/WriteBufferToFileSegment.cpp:137]: But `prepareVerticalMergeForAllColumns` is waiting for a `ReadBufferFromFile`, not a `ReadBufferFromEmptyFile`. It's going to still fail, right?

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@kssenii, is it possible to add a test?

**Comment by kssenii** [comment]
> @kssenii, is it possible to add a test?

Done in https://github.com/ClickHouse/ClickHouse/pull/65949/commits/2ce564daa0a7eb98b46dbed50d2eddfff8a731c2

**Comment by kssenii** [comment]
> Upgrade check (debug) 

Fixed in https://github.com/ClickHouse/ClickHouse/pull/65945

> Stateless tests (tsan) [5/5] 

`03144_parallel_alter_add_drop_column_zookeeper_on_steroids` -- flaky

> AST fuzzer (msan) 

https://github.com/ClickHouse/ClickHouse/issues/66027

**Pull request merged**